cltk/phonology/orthophonology.py

```diff
@@ -0,0 +1,643 @@
+		for feature_name, feature_value in features.items():
+			if not issubclass(feature_name, PhonologicalFeature):
+				raise TypeError(str(feature_name) + ' is not a phonological feature')
+			if type(feature_value) != feature_name:
+				raise TypeError(str(feature_value) + ' is not a ' + str(feature_name))
+
+		self.features = features
+		self.ipa = ipa
+
+	def is_vowel(self):
+		return self[Consonantal] == Consonantal.neg
+
+	def _check_disjunctive_features(self, feature_values):
+		return reduce(lambda a, b: a or b, [self[type(f)] == f for f in feature_values])
+
+	def merge(self, other):
+		phoneme = deepcopy(self)
+
+		# special case for list of phonemes
+		if type(other) == list and len(other) > 0 and issubclass(type(other[0]), AbstractPhoneme):
+			return other
+
+		if issubclass(type(other), AbstractPhoneme):
+			feature_values = other.features.values()
+		elif type(other) != list and type(other) != tuple:
+			feature_values = [other]
+		else:
+			feature_values = other
+
+		for f in feature_values:
+			if type(f) == list:
+				for inner_f in f:
+					phoneme[type(inner_f)] = inner_f
+			elif issubclass(type(f), AbstractPhoneme):
+				phoneme = phoneme << f
+			else:
+				phoneme[type(f)] = f
+
+		if issubclass(type(other), AbstractPhoneme) and other.ipa is not None:
+			phoneme.ipa = other.ipa
+
+		return phoneme
+
+	def is_equal(self, other):
+		return self.features == other.features
+
+	def matches(self, other):
+		if other is None:
+			return False
+		if type(other) == list or issubclass(type(other), PhonologicalFeature):
+			other = phoneme(other)
+		return other.features.items() >= self.features.items()
+
+	def __getitem__(self, feature_name):
+		'''
+		Use dict-type syntax for accessing the values of features.
+		'''
+		if not issubclass(feature_name, PhonologicalFeature):
+			raise TypeError(str(feature_name) + ' is not a phonological feature')
+		return self.features.get(feature_name, None)
+
+	def __setitem__(self, feature_name, feature_value):
+		if not issubclass(feature_name, PhonologicalFeature):
+			raise TypeError(str(feature_name) + ' is not a phonological feature')
+		if type(feature_value) != feature_name:
+			raise TypeError(str(feature_value) + ' is not a ' + str(feature_name))
+		self.features[feature_name] = feature_value
+		return self
+
+	def __str__(self):
+		return  ('IPA:{0} '.format(self.ipa) if self.ipa is not None else '') + \
+		' '.join([str(v) for v in self.features.values()])
+
+	__repr__ = __str__
+
+	def __eq__(self, other):
+		return self.is_equal(other)
+
+	def __le__ (self, other):
+		return self.matches(other)
+
+	def __ge__ (self, other):
+		return other.matches(self)
+
+	def __lt__ (self, other):
+		return other.is_more_sonorous(self)
+
+	def __gt__ (self, other):
+		return self.is_more_sonorous(other)
+
+	def __rshift__(self, other):
+		return PhonologicalRule(
+			condition = lambda _, target, __: self <= target,
+			action = lambda target : target << other)
+
+	def __lshift__(self, other):

```

Comment:
It looks like black magic for most people (yes, like C++ 😅), but it's ok

Reply:
Yeah, operator overloading in C++ is really annoying because obscure.  The first time I saw `cout << "hello, world'"` I felt manipulated :)  There's little defence, except to argue that using operator overloading one can put together DSLs in languages that lack macro facilities.

Reply:
I would much prefer real macros.  There is a project for adding macros to Python: https://github.com/lihaoyi/macropy  But I'm guessing that we may not want to create a dependency on a library that is this powerful and yet not part of the core distribution.